ntoskrnl/io/pnpmgr/devaction.c

```diff
@@ -2801,54 +2290,180 @@ IoRequestDeviceEject(IN PDEVICE_OBJECT PhysicalDeviceObject)
+                }
+                break;
+            case DeviceNodeStartPostWork:
+                DPRINT("DeviceNodeStartPostWork %wZ\n", &currentNode->InstancePath);
+                status = PiStartDeviceFinal(currentNode);
+                doProcessAgain = TRUE;
+                break;
+            case DeviceNodeStarted:
+                if (currentNode->Flags & DNF_REENUMERATE)
+                {
+                    DPRINT("DeviceNodeStarted REENUMERATE %wZ\n", &currentNode->InstancePath);
+                    currentNode->Flags &= ~DNF_REENUMERATE;
+                    status = PiIrpQueryDeviceRelations(currentNode, BusRelations);
 
-        if (NT_SUCCESS(Status))
-        {
-            /* Start the service and begin PnP initialization of the device again */
-            DPRINT("A new driver will be loaded for '%wZ' (no FDO above)\n", &DeviceNode->InstancePath);
-            Status = IopActionInitChildServices(DeviceNode, DeviceNode->Parent);
+                    // again, skip DeviceNodeEnumeratePending as with starting sequence
+                    PiSetDevNodeState(currentNode, DeviceNodeEnumerateCompletion);
+                    doProcessAgain = TRUE;
+                }
+                break;
+            case DeviceNodeQueryStopped:
+                // we're here after sending IRP_MN_QUERY_STOP_DEVICE
+                status = currentNode->CompletionStatus;
+                if (NT_SUCCESS(status))
+                {
+                    PiSetDevNodeState(currentNode, DeviceNodeStopped);
+                }
+                else
+                {
+                    PiIrpCancelStopDevice(currentNode);
+                    PiSetDevNodeState(currentNode, DeviceNodeStarted);
+                }
+                break;
+            case DeviceNodeStopped:
+                // do resource rebalance (not implemented)
+                ASSERT(FALSE);
+                break;
+            case DeviceNodeRestartCompletion:
+                break;
+            case DeviceNodeEnumeratePending: // skipped on XP/2003
+                break;
+            case DeviceNodeEnumerateCompletion:
+                DPRINT("DeviceNodeEnumerateCompletion %wZ\n", &currentNode->InstancePath);
+                status = PiEnumerateDevice(currentNode);
+                doProcessAgain = TRUE;
+                break;
+            case DeviceNodeAwaitingQueuedDeletion:
+                break;
+            case DeviceNodeAwaitingQueuedRemoval:
+                DPRINT("DeviceNodeAwaitingQueuedRemoval %wZ\n", &currentNode->InstancePath);
+                status = IopRemoveDevice(currentNode);
+                break;
+            case DeviceNodeQueryRemoved:
+                break;
+            case DeviceNodeRemovePendingCloses:
+                break;
+            case DeviceNodeRemoved:
+                break;
+            case DeviceNodeDeletePendingCloses:
+                break;
+            case DeviceNodeDeleted:
+                break;
```

Comment:
Are all these cases kept like that because they need to be implemented in the future?
If not maybe just doing:
```c
case A:
case B:
...
case N:
default:
    break:
```
would be clearer on the intention?